doc/source/serve/doc_code/vllm_openai_example.py

```diff
@@ -0,0 +1,115 @@
+        if engine_args.served_model_name is not None:
+            served_model_names = engine_args.served_model_name
+        else:
+            served_model_names = [engine_args.model]
+        self.openai_serving_chat = OpenAIServingChat(self.engine,
+                                                        served_model_names,
+                                                        response_role,
+                                                        lora_modules,
+                                                        chat_template)
+
+    # API Reference here - https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html
+    @app.post("/v1/chat/completions")
+    async def create_chat_completion(self,
+                                request: ChatCompletionRequest,
+                                raw_request: Request):
+        print("request", request)
+        print("raw request", raw_request)
+        generator = await self.openai_serving_chat.create_chat_completion(
+            request, raw_request)
+        if isinstance(generator, ErrorResponse):
+            return JSONResponse(content=generator.model_dump(),
+                                status_code=generator.code)
+        if request.stream:
+            return StreamingResponse(content=generator,
+                                    media_type="text/event-stream")
+        else:
+            assert isinstance(generator, ChatCompletionResponse)
+            return JSONResponse(content=generator.model_dump())
+
+def deployment(cli_args: Dict[str, str]) -> serve.Application:
+    """
+    Refer to https://docs.vllm.ai/en/latest/models/engine_args.html for arguments
+
+    """
+    parser = make_arg_parser()
+    arg_strings = []
+    for key, value in cli_args.items():
+        arg_strings.extend([f'--{key}', str(value)])
+    print(arg_strings)
+    parsed_args = parser.parse_args(args=arg_strings)
+    engine_args = AsyncEngineArgs.from_cli_args(parsed_args)
+    engine_args.worker_use_ray = True
+    tp = engine_args.tensor_parallel_size
+    pg_resources = []
+    print(f"Tensor parallelism = {tp}")
+    pg_resources.append({"CPU": 1}) # for the deployment
+    for i in range(tp):
+        pg_resources.append({"CPU": 2, "GPU": 1}) # for the vLLM workers
+    print("pg resources", pg_resources)
+    print("engine args", engine_args)
+    print("response role", parsed_args.response_role)
+    print("lora modules", parsed_args.lora_modules)
+    print("chat template", parsed_args.chat_template)
+    return VLLMDeployment.options(placement_group_bundles=pg_resources, placement_group_strategy="STRICT_PACK").bind(engine_args,
+        parsed_args.response_role, parsed_args.lora_modules, parsed_args.chat_template)
+
+# __serve_example_end__
+
+if __name__ == "__main__":
+    serve.run(deployment(cli_args = {"model": "NousResearch/Meta-Llama-3-8B-Instruct", "tensor-parallel-size": "2"}))
+    from openai import OpenAI
```

Comment:
We can use this part in the request example code. Just need to add another annotation like `# __request_example_start__` and `# __request_example_end__` around this block. Then `literalinclude` can pick it up again in the markdown file.